File: benchmark/qwen3-asr-webgpu/run_page.py
"""Drive headless Chrome at the spike page over CDP (for boxes without node).

usage: python run_page.py <chrome-binary> <url> [timeoutSec] [extra chrome flags...]
needs: pip install websocket-client
"""
import json
import os
import random
import subprocess
import sys
import tempfile
import time
import urllib.request

import websocket  # websocket-client

chrome, url = sys.argv[1], sys.argv[2]
timeout = int(sys.argv[3]) if len(sys.argv) > 3 else 900
extra = sys.argv[4:]
port = 9222 + random.randint(0, 500)
profile = os.path.join(tempfile.gettempdir(), f"spike-chrome-profile-{port}")
# NO_HEADLESS=1 opens a real window (needed on Linux boxes where headless Chrome only gets a
# SwiftShader adapter; run it on the box's own display, same as run_page.mjs).
headless = [] if os.environ.get("NO_HEADLESS") else ["--headless=new"]
flags = [chrome, *headless, "--no-sandbox", "--disable-dev-shm-usage", f"--remote-debugging-port={port}", f"--user-data-dir={profile}",
         "--enable-unsafe-webgpu", "--ignore-gpu-blocklist", "--window-size=800,600", *extra, "about:blank"]
proc = subprocess.Popen(flags, stdout=subprocess.DEVNULL, stderr=subprocess.PIPE)

# Chrome must not outlive this script: a dropped SSH session (SIGHUP) or a timeout kill used
# to leave a headless Chrome behind, holding the model in its GPU process for hours and
# contaminating the next run's memory measurements.
import atexit  # noqa: E402
import signal  # noqa: E402
atexit.register(lambda: proc.poll() is None and proc.kill())
for _sig in (signal.SIGHUP, signal.SIGTERM, signal.SIGINT):
    signal.signal(_sig, lambda *_: sys.exit(130))

ws_url = None
for _ in range(150):
    try:
        with urllib.request.urlopen(f"http://127.0.0.1:{port}/json", timeout=2) as r:
            for t in json.load(r):
                if t.get("type") == "page":
                    ws_url = t["webSocketDebuggerUrl"]
                    break
    except Exception:
        pass
    if ws_url:
        break
    time.sleep(0.2)
if not ws_url:
    print("no CDP target", proc.stderr.read()[-800:].decode(errors="replace"))
    proc.kill()
    sys.exit(3)

ws = websocket.create_connection(ws_url, suppress_origin=True)
ws.settimeout(5)
mid = 0


def send(method, params=None):
    global mid
    mid += 1
    ws.send(json.dumps({"id": mid, "method": method, "params": params or {}}))
    while True:
        msg = json.loads(ws.recv())
        handle_event(msg)
        if msg.get("id") == mid:
            return msg


def handle_event(msg):
    if msg.get("method") == "Runtime.consoleAPICalled":
        text = " ".join(str(a.get("value", a.get("description", ""))) for a in msg["params"]["args"])
        if text.startswith(("RESULT ", "STATUS ")):
            print(text, flush=True)
        elif msg["params"]["type"] in ("error", "warning"):
            print("[console." + msg["params"]["type"] + "] " + text[:500], flush=True)
    elif msg.get("method") == "Runtime.exceptionThrown":
        print("[exception] " + json.dumps(msg["params"]["exceptionDetails"])[:800], flush=True)


send("Runtime.enable")
send("Page.enable")
send("Page.navigate", {"url": url})
start = time.time()
done = False
while time.time() - start < timeout:
    # drain events for ~1s
    t_end = time.time() + 1.0
    while time.time() < t_end:
        try:
            ws.settimeout(max(0.05, t_end - time.time()))
            handle_event(json.loads(ws.recv()))
        except websocket.WebSocketTimeoutException:
            break
    ws.settimeout(5)
    r = send("Runtime.evaluate", {"expression": "JSON.stringify({done: window.__result?.done})", "returnByValue": True})
    v = r.get("result", {}).get("result", {}).get("value")
    if v and json.loads(v).get("done"):
        done = True
        break
if not done:
    print(f"TIMEOUT after {timeout}s")
fin = send("Runtime.evaluate", {"expression": "JSON.stringify(window.__result)", "returnByValue": True})
print("FINAL " + str(fin.get("result", {}).get("result", {}).get("value"))[:20000])
ws.close()
proc.terminate()
sys.exit(0 if done else 2)

File: benchmark/qwen3-asr-webgpu/run_page.mjs
// Drive a headless Chromium at the spike page over CDP and print RESULT lines.
// usage: node run_page.mjs <chrome-binary> <url> [timeoutSec] [extra chrome flags...]
// env: PROFILE_DIR (prefix for --user-data-dir), SHOW_GPU_LOG=1 (print GPU-related stderr lines)
// Works on Node >= 22 without dependencies (global WebSocket); falls back to the `ws` package.
import { spawn } from 'node:child_process';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

let WS = globalThis.WebSocket;
if (!WS) {
  const { createRequire } = await import('node:module');
  WS = createRequire('/home/jiangzhuo/Desktop/kizunaai/sokuji/node_modules/')('ws');
}

const [chrome, url, timeoutArg, ...extra] = process.argv.slice(2);
const timeoutMs = (parseInt(timeoutArg || '900', 10)) * 1000;
const port = 9222 + Math.floor(Math.random() * 500);
const profile = `${process.env.PROFILE_DIR || join(tmpdir(), 'spike-chrome-profile')}-${port}`;
const linuxGpuFlags = process.platform === 'linux'
  ? ['--enable-features=Vulkan,WebGPU', '--use-angle=vulkan', '--enable-dawn-features=allow_unsafe_apis']
  : [];
const headless = process.env.NO_HEADLESS ? [] : ['--headless=new'];
const flags = [...headless, '--no-sandbox', '--disable-dev-shm-usage', `--remote-debugging-port=${port}`, `--user-data-dir=${profile}`,
  '--enable-unsafe-webgpu', '--ignore-gpu-blocklist', ...linuxGpuFlags, '--window-size=800,600', ...extra, 'about:blank'];
const proc = spawn(chrome, flags, { stdio: ['ignore', 'pipe', 'pipe'] });
// Chrome must not outlive this script: a dropped SSH session (SIGHUP) or a timeout kill used to
// leave a headless Chrome behind, holding the model in its GPU process and contaminating the
// next run's memory measurements.
process.on('exit', () => { if (proc.exitCode === null) proc.kill('SIGKILL'); });
for (const sig of ['SIGHUP', 'SIGTERM', 'SIGINT']) process.on(sig, () => process.exit(130));
let stderr = '';
proc.stderr.on('data', (d) => { stderr += d; });
const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function waitForTarget() {
  for (let i = 0; i < 150; i++) {
    try {
      const res = await fetch(`http://127.0.0.1:${port}/json`);
      const list = await res.json();
      const page = list.find((t) => t.type === 'page');
      if (page) return page.webSocketDebuggerUrl;
    } catch {}
    await sleep(200);
  }
  throw new Error('no CDP target; stderr tail: ' + stderr.slice(-800));
}

const wsUrl = await waitForTarget();
const ws = new WS(wsUrl);
await new Promise((r, j) => { ws.addEventListener('open', r); ws.addEventListener('error', j); });
let id = 0; const pending = new Map();
ws.addEventListener('message', (ev) => {
  const msg = JSON.parse(typeof ev.data === 'string' ? ev.data : ev.data.toString());
  if (msg.id && pending.has(msg.id)) { pending.get(msg.id)(msg); pending.delete(msg.id); }
  else if (msg.method === 'Runtime.consoleAPICalled') {
    const text = msg.params.args.map((a) => a.value ?? a.description ?? '').join(' ');
    if (text.startsWith('RESULT ') || text.startsWith('STATUS ')) console.log(text);
    else if (msg.params.type === 'error' || msg.params.type === 'warning') console.log('[console.' + msg.params.type + '] ' + text.slice(0, 500));
  } else if (msg.method === 'Runtime.exceptionThrown') {
    console.log('[exception] ' + JSON.stringify(msg.params.exceptionDetails).slice(0, 800));
  }
});
const send = (method, params = {}) => new Promise((res) => { const i = ++id; pending.set(i, res); ws.send(JSON.stringify({ id: i, method, params })); });
await send('Runtime.enable');
await send('Page.enable');
await send('Page.navigate', { url });
const start = Date.now();
let done = false;
while (Date.now() - start < timeoutMs) {
  await sleep(1000);
  const r = await send('Runtime.evaluate', { expression: 'JSON.stringify({done: window.__result?.done, n: window.__result?.clips?.length})', returnByValue: true });
  const v = r.result?.result?.value ? JSON.parse(r.result.result.value) : null;
  if (v?.done) { done = true; break; }
}
if (!done) console.log('TIMEOUT after ' + timeoutMs / 1000 + 's');
if (process.env.SHOW_GPU_LOG) console.log('[stderr gpu lines]\n' + stderr.split('\n').filter((l) => /vulkan|dawn|webgpu|gpu|angle/i.test(l)).slice(-25).join('\n'));
const fin = await send('Runtime.evaluate', { expression: 'JSON.stringify(window.__result)', returnByValue: true });
console.log('FINAL ' + (fin.result?.result?.value || 'null').slice(0, 20000));
ws.close();
proc.kill('SIGTERM');
await sleep(500);
process.exit(done ? 0 : 2);
